Objects/typeobject.c

```diff
@@ -5678,8 +5678,13 @@ is_dunder_name(PyObject *name)
 static PyObject *
 update_cache(struct type_cache_entry *entry, PyObject *name, unsigned int version_tag, PyObject *value)
 {
-    _Py_atomic_store_uint32_relaxed(&entry->version, version_tag);
+    // We must write the value first to avoid _Py_TryXGetStackRef()
+    // operating on an invalid (already deallocated) value inside
+    // _PyType_LookupRefAndVersion().  If we write the version first then a
+    // reader could pass the "entry_version == type_version" check but could
+    // be using the old entry value.
     _Py_atomic_store_ptr_relaxed(&entry->value, value); /* borrowed */
+    _Py_atomic_store_uint32_relaxed(&entry->version, version_tag);
```

Comment:
If the ordering here is important, then I think we need at least:

`_Py_atomic_store_uint32_release(&entry->version, version_tag)`

To ensure that the write to `entry->value` is visible before the write to `entry->version`.

Reply:
I'd probably also move the write to `entry->name` up. I'm not really sure if it matters, but it seems cleaner to me to have the `version_tag` written last.